### ISSUE TITLE
Long Startup Time, /nextcloud Permissions

### ISSUE THREAD

**Issue by PhasecoreX** [benyanke/docker-nextcloud#13]
Every time I recreate my Nextcloud container, it goes through the process of making sure all permissions are correct for a bunch of folders. This is great and all, except that when it processes the /nextcloud folder, it always needs to correct the permissions. Since this is such a large directory, it takes several minutes (of downtime and disk thrashing) to process.

Doing a bit of research, it looks like the user and group of these files is "nobody" and "nogroup", respectively. This probably comes from building the docker file when the whole thing is un-tarred.

Does this folder need to have its permissions changed to still function? I tried turning off permission checking with the PERMISSION_RESET=0 environment variable, but then the /bin/s6-svscan /etc/s6.d command complains about permissions. I'd prefer permission checking to be on, but I also like my disks to not die doing the same task over and over.

**Comment by benyanke** [comment]
I'll have to work on this one. _In theory_ once the permissions are correct, there's no reason they'd need to constantly be re-reset. However, I haven't done quite as much testing with this workflow as I'd like, especially since in my main production install, I just allow the permissions to be reset every time.

**Comment by benyanke** [comment]
Could you provide a few more details about how you're starting it when you get the permissions complaints? A docker compose file or docker run script or something?

**Comment by benyanke** [comment]
At first glance at the code (https://github.com/benyanke/docker-nextcloud/commit/81ab13dd86ec216f7eed0ca39ce3b255bfa9833d), I think a useful solution might be to separate permission resets for data out from permission resets for the rest of the container. For example, there's not really a reason to stop permission fixing on /php or the log dir if the user is just trying to stop from resetting permissions on his 3 million file data dir.

Just stream of consciousness here - will continue to dig, and invite any further ideas or PRs you might have on the topic as well.

**Comment by PhasecoreX** [comment]
I'm using docker-compose to launch it along with a reverse proxy and database (not shown), like so:

```
services:
  web-nextcloud:
    image: benyanke/nextcloud:latest
    container_name: web-nextcloud
    restart: always
    volumes:
      - ./nextcloud/apps:/apps2
      - ./nextcloud/config:/config
      - ./nextcloud/data:/data
      - ./nextcloud/themes:/nextcloud/themes
    environment:
      - VIRTUAL_HOST=<url>
      - LETSENCRYPT_HOST=<url>
      - LETSENCRYPT_EMAIL=<email>
      - UID=1000
      - GID=1000
      - PERMISSION_RESET=0
      - UPLOAD_MAX_SIZE=10G
      - APC_SHM_SIZE=128M
      - OPCACHE_MEM_SIZE=128
      - CRON_PERIOD=15m
      - TZ="America/Detroit"
      - DOMAIN=<url>
      - DB_TYPE=mysql
    networks:
      - default
      - database
```

Pretty much exactly the same as the example ones in the readme.md. I just do a docker-compose up. The error I run into when running it this way:

```
Not updating permissions since $PERMISSION_RESET was not '1'
Cannot write into "config" directory
This can usually be fixed by giving the webserver write access to the config directory. See https://docs.nextcloud.com/server/13/go.php?to=admin-dir_permissions

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /nextcloud/lib/private/Console/Application.php:148
Stack trace:
#0 /nextcloud/console.php(89): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /nextcloud/occ(11): require_once('/nextcloud/cons...')
s6-svscan: fatal: unable to mkfifo .s6-svscan/control: Permission denied
#2 {main}
```

A couple of thoughts I have is that permissions shouldn't change too much, and we are already scanning over every file for a changed user/group. Thus, it might make more sense to have the find command (which is already checking every file) also do the chmod with -exec. The first pass might take a much longer time, but subsequent passes (at least over /data) will be much quicker, as one or two files needing a permissions change won't do a chmod -R over the whole directory.

The /nextcloud directory though? Not sure yet what to do about that. Seems like the permissions will always be incorrect, at least how I'm starting the container. I do agree that the volumes should have their permissions checked based on the PERMISSION_RESET environment variable, separate from the other directories being currently checked.

**Comment by benyanke** [comment]
Created https://github.com/benyanke/docker-nextcloud/tree/feature/permission-issues to track some solutions for this issue. Pushing to docker tag `feature-permission-issues`

**Comment by benyanke** [comment]
Any luck trying this? I'm thinking it should resolve your issue, and I'd love to merge this in, assuming it doesn't cause any issues.

**Comment by benyanke** [comment]
Excellent - I'll merge that in and continue.

Thanks for the tip on the official image. That's fascinating, and could be a good model for us here. I'll continue to look into that one.

I've long wanted to speed up startup of the container, thanks for the help moving that forward.

**Comment by GreenBlood** [comment]
Hey @benyanke I just tested around quickly @PhasecoreX 's idea with rsync and I came up with this https://github.com/GreenBlood/docker-nextcloud/tree/rsync-test

It works in two stages, first the docker build downloads and extracts the nextcloud tar in /nextcloudtemp
Then in the run.sh it rsyncs it to /nextcloud with the --chown $UID:$GID option of rsync. Since rsync is much faster the start-up time is really quick. (It deletes in the background /nextcloudtemp afterwards)
I moved the config and apps symlink after this rsync, also.

I have not tested it much but my setup is currently using this custom image and so far no error. The workflow seems fine to me but I may have missed something.
